properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+          in the input dictionary.
+        * Any other keyword arguments will be passed through to the Property
+          deserializers.
+
+        .. note::
+
+            UidModel instances are constructed with no input arguments
+            (ie :code:`cls()` is called). This means deserialization will
+            fail if the init method has been overridden to require
+            input parameters.
+        """
+        registry = kwargs.pop('registry', None)
+        if registry is None:
+            if not isinstance(value, dict):
+                raise ValueError('HasUID must deserialize from dictionary')
+            registry = value.copy()
+            uid = kwargs.get('uid', registry.get('__uid__'))
+        else:
+            uid = value
+        if uid in cls._INSTANCES and uid not in registry:
+            return cls._INSTANCES[uid]
+        elif uid in cls._INSTANCES:
+            raise ValueError('Uid already used: {}'.format(uid))
+        elif uid not in registry:
+            raise ValueError('Invalid uid: {}'.format(uid))
+        value = registry[uid]
+        if not isinstance(value, HasUID):
+            try:
+                input_class = value.get('__class__')
+            except AttributeError:
+                input_class = None
+            new_cls = cls._deserialize_class(input_class, trusted, strict)
+            new_inst = new_cls()
+            registry.update({uid: new_inst})
+            super(HasUID, cls).deserialize(
+                value=value,
+                trusted=trusted,
+                strict=strict,
+                registry=registry,
+                _instance=new_inst,
+                **kwargs
+            )
+        cls._INSTANCES[uid] = registry[uid]
+        return registry[uid]
+
+
+class Pointer(base.Instance):
+    """Property for HasUID instances where string UID pointer may be used
+
+    **Available keywords** (in addition to those inherited from
+    :ref:`Instance <instance>`):
+
+    * **enforce_uid** - Require Pointer strings to resolve into instances.
+      If False, the default, the Pointer property may be set to an arbitrary
+      string; if True, the string must correspond to an existing UID.
+    * **uid_prop** - Property name of the UID property on instance_class.
+      The default is 'uid'.
+    """
+
+    class_info = 'an instance or uid of an instance'
+
+    @property
+    def enforce_uid(self):
+        """Require Pointer strings to resolve into instances
+
+        If False, the default, the Pointer may be set to an arbitrary string;
+        if True, the string must correspond to an existing UID.
+        """
+        return getattr(self, '_enforce_uid', False)
+
+    @enforce_uid.setter
+    def enforce_uid(self, value):
+        self._enforce_uid = bool(value)
+
+    @property
+    def uid_prop(self):
+        """Property name of the UID property on instance_class
+
+        The default is 'uid'
+        """
+        return getattr(self, '_uid_prop', 'uid')
```

Comment:
Oh, interesting. How flexible is this? I guess I need to play with it a bit.

Reply:
I tried to make it somewhat flexible, but still ended up with requiring the uid to be a string. In theory, it could be like an int or something, but that becomes a bit complicated when `Instance` and `Pointer` validation start to overlap.

It should work fine if `uid_prop` isn't a `property` though...